File: frontend/src/smc/lib/disease-surveillance/detection.ts
import { endOfWeek, format, startOfWeek, subWeeks } from "date-fns";

export type DetectionRiskLevel = "LOW" | "MEDIUM" | "HIGH";

export type SurveillanceDetectionInput = {
  wardNumber: number;
  wardName: string;
  diseaseId: string;
  diseaseName: string;
  weeklyCounts: number[];
};

export type SurveillanceDetectionResult = {
  wardNumber: number;
  wardName: string;
  diseaseId: string;
  diseaseName: string;
  currentWeekCases: number;
  previousWeekCases: number;
  growthRate: number;
  growthTrend: "Increase" | "Decrease" | "Stable";
  riskLevel: DetectionRiskLevel;
  earlyOutbreakSignal: boolean;
  ruleTriggered: boolean;
  trendTriggered: boolean;
  anomalyTriggered: boolean;
  abnormalityScore: number;
  weeklyHistory: { label: string; cases: number }[];
  alertMessage: string;
  citizenPreview: string;
  explanation: string;
};

function round(value: number, digits = 1) {
  const factor = 10 ** digits;
  return Math.round(value * factor) / factor;
}

function getGrowthRate(currentWeekCases: number, previousWeekCases: number) {
  if (previousWeekCases <= 0) {
    return currentWeekCases > 0 ? 100 : 0;
  }

  return round(((currentWeekCases - previousWeekCases) / previousWeekCases) * 100);
}

function getGrowthTrend(growthRate: number): SurveillanceDetectionResult["growthTrend"] {
  if (growthRate > 5) {
    return "Increase";
  }

  if (growthRate < -5) {
    return "Decrease";
  }

  return "Stable";
}

function average(values: number[]) {
  if (values.length === 0) {
    return 0;
  }

  return values.reduce((sum, value) => sum + value, 0) / values.length;
}

function standardDeviation(values: number[]) {
  if (values.length === 0) {
    return 0;
  }

  const mean = average(values);
  const variance =
    values.reduce((sum, value) => sum + (value - mean) ** 2, 0) / values.length;

  return Math.sqrt(variance);
}

function buildAlertMessage(diseaseName: string, wardNumber: number) {
  return `Early warning: Possible outbreak of ${diseaseName} in Ward ${wardNumber}`;
}

function buildCitizenPreview(diseaseName: string, wardNumber: number) {
  const disease = diseaseName.toLowerCase();

  if (disease.includes("dengue")) {
    return `Dengue cases are increasing in your area (Ward ${wardNumber}). Please remove stagnant water, use mosquito protection, and seek medical advice if fever symptoms appear.`;
  }

  if (disease.includes("flu") || disease.includes("influenza")) {
    return `Flu cases are increasing in your area (Ward ${wardNumber}). Please practice hand hygiene, wear a mask if unwell, and visit a clinic if symptoms worsen.`;
  }

  if (disease.includes("diarr")) {
    return `Diarrheal illness is increasing in your area (Ward ${wardNumber}). Please drink safe water, maintain food hygiene, and seek care if dehydration symptoms appear.`;
  }

  return `${diseaseName} cases are increasing in your area (Ward ${wardNumber}). Please follow local health precautions, maintain hygiene, and contact a nearby health facility if symptoms appear.`;
}

export function buildWeeklyHistory(weeklyCounts: number[]) {
  const historyStart = startOfWeek(subWeeks(new Date(), weeklyCounts.length - 1), {
    weekStartsOn: 1,
  });

  return weeklyCounts.map((count, index) => ({
    label: format(endOfWeek(subWeeks(historyStart, -index), { weekStartsOn: 1 }), "dd MMM"),
    cases: count,
  }));
}

export function evaluateOutbreakSignals(
  input: SurveillanceDetectionInput,
): SurveillanceDetectionResult {
  const currentWeekCases = input.weeklyCounts.at(-1) ?? 0;
  const previousWeekCases = input.weeklyCounts.at(-2) ?? 0;
  const growthRate = getGrowthRate(currentWeekCases, previousWeekCases);
  const growthTrend = getGrowthTrend(growthRate);

  const baseline = input.weeklyCounts.slice(0, -1);
  const recentWindow = input.weeklyCounts.slice(-4);
  const mean = average(baseline);
  const deviation = standardDeviation(baseline);

  const ruleTriggered =
    (currentWeekCases >= 12 && growthRate >= 50) ||
    (currentWeekCases >= 8 && previousWeekCases === 0) ||
    (currentWeekCases >= 18 && growthRate >= 25);

  const mediumRule =
    !ruleTriggered &&
    ((currentWeekCases >= 8 && growthRate >= 20) ||
      (currentWeekCases >= 6 && growthRate >= 35));

  const trendTriggered =
    recentWindow.length >= 4 &&
    recentWindow[0] < recentWindow[1] &&
    recentWindow[1] < recentWindow[2] &&
    recentWindow[2] < recentWindow[3] &&
    recentWindow[3] >= 6;

  const anomalyThreshold = mean + Math.max(2 * deviation, 3);
  const anomalyTriggered =
    baseline.length >= 3 &&
    currentWeekCases >= 6 &&
    currentWeekCases > anomalyThreshold;

  const riskLevel: DetectionRiskLevel = ruleTriggered
    ? "HIGH"
    : mediumRule || trendTriggered || anomalyTriggered
      ? "MEDIUM"
      : "LOW";

  const earlyOutbreakSignal = ruleTriggered || trendTriggered || anomalyTriggered;
  const abnormalityScore = round(
    (ruleTriggered ? 45 : 0) +
      (mediumRule ? 20 : 0) +
      (trendTriggered ? 20 : 0) +
      (anomalyTriggered ? 25 : 0) +
      Math.min(10, currentWeekCases),
    1,
  );

  const triggers = [
    ruleTriggered ? "high case load and rapid weekly growth" : null,
    trendTriggered ? "continuous increase over recent weeks" : null,
    anomalyTriggered ? "statistical spike above baseline" : null,
  ].filter(Boolean);

  const explanation =
    triggers.length > 0
      ? `Flagged due to ${triggers.join(", ")}.`
      : "Current ward-level counts remain within the normal recent range.";

  return {
    wardNumber: input.wardNumber,
    wardName: input.wardName,
    diseaseId: input.diseaseId,
    diseaseName: input.diseaseName,
    currentWeekCases,
    previousWeekCases,
    growthRate,
    growthTrend,
    riskLevel,
    earlyOutbreakSignal,
    ruleTriggered,
    trendTriggered,
    anomalyTriggered,
    abnormalityScore,
    weeklyHistory: buildWeeklyHistory(input.weeklyCounts),
    alertMessage: buildAlertMessage(input.diseaseName, input.wardNumber),
    citizenPreview: buildCitizenPreview(input.diseaseName, input.wardNumber),
    explanation,
  };
}

File: backend/app/modules/ai/detection.py
import math
from typing import List, Dict, Any


def round_val(value: float, digits: int = 1) -> float:
    factor = 10 ** digits
    return round(value * factor) / factor


def get_growth_rate(current_cases: int, previous_cases: int) -> float:
    if previous_cases <= 0:
        return 100.0 if current_cases > 0 else 0.0
    return round_val(((current_cases - previous_cases) / previous_cases) * 100.0)


def average(values: List[int]) -> float:
    if not values:
        return 0.0
    return sum(values) / len(values)


def standard_deviation(values: List[int]) -> float:
    if not values:
        return 0.0
    avg = average(values)
    variance = sum((x - avg) ** 2 for x in values) / len(values)
    return math.sqrt(variance)


def evaluate_outbreak_signals(
    ward_number: int,
    ward_name: str,
    disease_id: str,
    disease_name: str,
    weekly_counts: List[int]
) -> Dict[str, Any]:
    current_cases = weekly_counts[-1] if weekly_counts else 0
    previous_cases = weekly_counts[-2] if len(weekly_counts) >= 2 else 0
    growth_rate = get_growth_rate(current_cases, previous_cases)

    growth_trend = "Stable"
    if growth_rate > 5:
        growth_trend = "Increase"
    elif growth_rate < -5:
        growth_trend = "Decrease"

    baseline = weekly_counts[:-1] if len(weekly_counts) > 1 else [0]
    recent_window = weekly_counts[-4:] if len(weekly_counts) >= 4 else weekly_counts
    mean_val = average(baseline)
    dev_val = standard_deviation(baseline)

    rule_triggered = (
        (current_cases >= 12 and growth_rate >= 50) or
        (current_cases >= 8 and previous_cases == 0) or
        (current_cases >= 18 and growth_rate >= 25)
    )

    medium_rule = (
        not rule_triggered and (
            (current_cases >= 8 and growth_rate >= 20) or
            (current_cases >= 6 and growth_rate >= 35)
        )
    )

    trend_triggered = (
        len(recent_window) >= 4 and
        recent_window[0] < recent_window[1] < recent_window[2] < recent_window[3] and
        recent_window[3] >= 6
    )

    anomaly_threshold = mean_val + max(2 * dev_val, 3.0)
    anomaly_triggered = (
        len(baseline) >= 3 and
        current_cases >= 6 and
        current_cases > anomaly_threshold
    )

    if rule_triggered:
        risk_level = "HIGH"
    elif medium_rule or trend_triggered or anomaly_triggered:
        risk_level = "MEDIUM"
    else:
        risk_level = "LOW"

    early_outbreak_signal = rule_triggered or trend_triggered or anomaly_triggered
    abnormality_score = round_val(
        (45.0 if rule_triggered else 0.0) +
        (20.0 if medium_rule else 0.0) +
        (20.0 if trend_triggered else 0.0) +
        (25.0 if anomaly_triggered else 0.0) +
        min(10.0, float(current_cases)),
        1
    )

    triggers = []
    if rule_triggered:
        triggers.append("high case load and rapid weekly growth")
    if trend_triggered:
        triggers.append("continuous increase over recent weeks")
    if anomaly_triggered:
        triggers.append("statistical spike above baseline")

    explanation = (
        f"Flagged due to {', '.join(triggers)}."
        if triggers else
        "Current ward-level counts remain within the normal recent range."
    )

    return {
        "wardNumber": ward_number,
        "wardName": ward_name,
        "diseaseId": disease_id,
        "diseaseName": disease_name,
        "currentWeekCases": current_cases,
        "previousWeekCases": previous_cases,
        "growthRate": growth_rate,
        "growthTrend": growth_trend,
        "riskLevel": risk_level,
        "earlyOutbreakSignal": early_outbreak_signal,
        "ruleTriggered": rule_triggered,
        "trendTriggered": trend_triggered,
        "anomalyTriggered": anomaly_triggered,
        "abnormalityScore": abnormality_score,
        "triggers": triggers,
        "explanation": explanation
    }
